server/src/main/java/org/opensearch/indices/NodeIndicesStats.java

```diff
@@ -74,23 +76,22 @@
  */
 @PublicApi(since = "1.0.0")
 public class NodeIndicesStats implements Writeable, ToXContentFragment {
-    private CommonStats stats;
-    private Map<Index, List<IndexShardStats>> statsByShard;
+    protected CommonStats stats;
+    protected Map<Index, CommonStats> statsByIndex;
+    protected Map<Index, List<IndexShardStats>> statsByShard;
 
     public NodeIndicesStats(StreamInput in) throws IOException {
         stats = new CommonStats(in);
+        if (in.getVersion().onOrAfter(Version.V_3_0_0)) {
+            // contains statsByIndex
+            if (in.readBoolean()) {
+                statsByIndex = new HashMap<>();
+                readStatsByIndex(in);
+            }
+        }
         if (in.readBoolean()) {
-            int entries = in.readVInt();
             statsByShard = new HashMap<>();
-            for (int i = 0; i < entries; i++) {
-                Index index = new Index(in);
-                int indexShardListSize = in.readVInt();
-                List<IndexShardStats> indexShardStats = new ArrayList<>(indexShardListSize);
-                for (int j = 0; j < indexShardListSize; j++) {
-                    indexShardStats.add(new IndexShardStats(in));
-                }
-                statsByShard.put(index, indexShardStats);
-            }
+            readStatsByShards(in);
```

Comment:
`readStatsByShards()` is just reading the same statsByShard as it does normally in other cases and previous versions. Made another function for improved code readability and simplicity.